src/inmanta/server/agentmanager.py

```diff
@@ -232,44 +268,62 @@ def get_agent_client(self, tid: uuid.UUID, endpoint: str) -> Optional[ReturnClie
             sid = session.get_id()
 
             LOGGER.debug("expiring session %s", sid)
-
             del self.sessions[sid]
 
-            env = await data.Environment.get_by_id(tid)
-            if env is None:
-                LOGGER.warning("The environment id %s, for agent %s does not exist!", tid, sid)
-
-            aps = await data.AgentProcess.get_by_sid(sid=sid)
-            if aps is None:
-                LOGGER.info("expiring session on none existant process sid:%s", sid)
-            else:
-                await aps.update_fields(expired=now)
+            endpoints_with_new_primary: List[Tuple[str, Optional[protocol.Session]]] = []
+            for endpoint in session.endpoint_names:
+                key = (tid, endpoint)
+                current_active_session = self.tid_endpoint_to_session.get(key, None)
+                if current_active_session is not None and current_active_session == session:
+
+                    new_active_session = self._get_session_for(tid, endpoint)
+                    if new_active_session is not None:
+                        self.tid_endpoint_to_session[key] = new_active_session
+                        self.add_background_task(new_active_session.get_client().set_state(endpoint, True))
+                        endpoints_with_new_primary.append((endpoint, new_active_session))
+                    else:
+                        del self.tid_endpoint_to_session[key]
+                        endpoints_with_new_primary.append((endpoint, None))
+
+            await self._log_session_expiry_to_db(tid, endpoints_with_new_primary, session, now)
```

Comment:
Should this be a background task? Now it is under the session lock, blocking all other sessions.